### PR TITLE
add new MetaMetricsController

### PR DESCRIPTION
**Pull request by brad-decker**
### Adds new MetaMetricsController
* Moves most constants previously declared in `shared/modules/metametrics` to `shared/constants/metametrics`
* Moves all global type declarations to `shared/constants/metametrics`
* Changes the `segmentNoop` into a factory that creates a mocked module. This new mocked module should behave closer to the way that Segment's track method actually works (including queueing). This is helpful for testing purposes and means that e2e tests _should_ timeout if we accidentally wait for events without `flushImmediately` set to true. 
* Moves as many event/page tracking implementation details from the UI into a new MetaMetrics controller.
* Moves and migrates `metaMetrics` related fields from preferences controller to MetaMetrics controller.
* No longer required to have a factory for generating `trackMetaMetricsEvent` functions. As many details as possible are tracked in the new controller's store instead of having to be provided by the consumer of `.trackEvent.`
* Also adds `.getNetworkIdentifier` method on the network controller as it didn't' really make sense to have in the Metametrics controller.
* Adds 95%+ code coverage on the new controller, where previously the MetaMetrics implementation was untested

### Rationale
As we do not have great control over performing operations before the extension UI completes exit/close events, we cannot rely on the provided segment utilities for flushing the event queue. We also have seen that even with batching disabled (setting flushAt: 1), some event loss is still occurring. By having the background process handle communication with Segment's API, we can use the batching mechanism of Segment and capture events from short-lived UI flows.

### CODE REVIEW

**Comment by brad-decker** [comment]
@rekmarks and @Gudahtt not ready for a full review, but could you check and see if I'm doing anything just horrific with how I've set up this new controller? Any tips, pointers, help is appreciated -- it is working in its current state but lots to cleanup, and would like to get some tests written for the controller.

**Review by Gudahtt — COMMENTED** [review]
This general approach seems good to me! It does seem appropriate for Segment to have its own controller, since there is associated state (I see that the state hasn't been moved into the controller yet, but that can be in a follow up PR).

I've left a couple of initial comments but I haven't looked closely at most of this yet

File: app/scripts/controllers/metametrics.js [hunk 1]
```diff
@@ -0,0 +1,409 @@
...
+)
+
+// This allows us to overwrite the metric destination for testing purposes
+const host = process.env.SEGMENT_HOST ?? undefined
+
+// We do not want to track events on development builds unless specifically
+// provided a SEGMENT_WRITE_KEY. This also holds true for test environments and
+// E2E, which is handled in the build process by never providing the SEGMENT_WRITE_KEY
+// when process.env.IN_TEST is truthy
+export const segment =
```
Comment (Gudahtt) [inline app/scripts/controllers/metametrics.js:77]: I'd suggest that both of these get moved into the constructor. It'll be easier to test that way - then all of these `process.env` variables can be passed in as constructor options, so we don't have to mutate `process.env` in the tests.

Comment (Gudahtt) [reply app/scripts/controllers/metametrics.js:77]: Also it's better to avoid stateful things being in module scope, because then we end up having shared state in between unit tests.

File: app/scripts/controllers/metametrics.js [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+   * @returns {Promise<void>}
+   */
+  trackEvent(payload, options) {
+    // event and category are required fields for all payloads
+    if (!payload.event || !payload.category) {
+      throw new Error('Must specify event and category.')
+    }
+
+    if (!this.participateInMetaMetrics && !options.isOptIn) {
+      return Promise.resolve()
```
Comment (Gudahtt) [inline app/scripts/controllers/metametrics.js:346]: Better to make this function `async` if it's doing async things. Right now it has an inconsistent return value in the event of failure - it'll throw an Error directly in some cases, but it'll return a rejected Promise in others.

File: ui/app/store/actions.js [hunk 3]
```diff
@@ -2733,3 +2738,77 @@ export function getCurrentWindowTab() {
...
+      }
+      resolve()
+    })
+  })
+}
+
+let previousMatch
+
+export function trackPage(location, environmentType, page, referrer) {
+  return async (_, getState) => {
```
[the remark this answers is not in the document]
Comment (brad-decker) [reply ui/app/store/actions.js:2750]: it was originally all in the `useEffect` in the metametrics context, I can move it back there. I thought I would try and lift some of the complexities out of react and into the action, but I agree with you. Also, this now includes some tests. more to come. thanks for your feedback @Gudahtt

File: test/unit/app/controllers/metametrics-test.js [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+    it('should immediately flush queue if flushImmediately set to true', async function () {
+      const metaMetricsController = getMetaMetricsController()
+      const value = await metaMetricsController.trackEvent(
+        {
+          event: 'Fake Event',
+          category: 'Unit Test',
+        },
+        { flushImmediately: true },
+      )
+      assert.strictEqual(value, undefined)
```
Comment (brad-decker) [inline test/unit/app/controllers/metametrics-test.js:141]: wasn't sure how to test that a promise resolves in < certain ms, so this could be improved.

File: ui/app/store/actions.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (Gudahtt) [reply ui/app/store/actions.js]: Yeah, I understand the impulse. Generally less logic in the UI is better, but perhaps not at the cost of creating a "fake" thunk.

The idea place for his would be in our routing slice, if we handled routing in Redux that is. That's something I wanted to do someday.

File: test/unit/app/controllers/metametrics-test.js (hunk 4 above)
Comment (Gudahtt) [reply test/unit/app/controllers/metametrics-test.js:287]: I created a utility to help with this, in `test/lib/wait-until-called.js`. You can see how it's used in the incoming transaction controller tests.

File: development/mock-segment.js [hunk 5]
```diff
@@ -8,7 +8,7 @@ function onRequest(request, response, events) {
   console.log(`${prefix}: ${request.method} ${request.url}`)
   const eventDescriptions = events.map((event) => {
     if (event.type === 'track') {
-      return event.event
+      return `${event.event}, ${JSON.stringify(event.properties)}`
```
Comment (brad-decker) [inline development/mock-segment.js:11]: i'll revert this before moving from draft.

File: shared/constants/metametrics.js [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+ *  segment source that marks the event data as not conforming to our schema
+ * @property {number} metaMetricsSendCount - the count of the number of sends
+ *  that have triggered a metrics event used to determine if we should allow
+ *  the user's id to be included in the payload periodically
+ */
+
+export const METAMETRICS_ANONYMOUS_ID = '0x0000000000000000'
+
+/**
+ * This object is used
```
Comment (brad-decker) [inline shared/constants/metametrics.js:78]: it is, i swear. I'll work on a few documentation things I've already found -- but should still be good for code review.

File: shared/constants/metametrics.js [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+// intent of having a value less than 10 seconds.
+export const SEGMENT_FLUSH_INTERVAL = 5000
+
+/**
+ * Creates a mock segment module for usage in test environments. This is used
+ * when building the application in test mode to catch event calls and prevent
+ * them from being sent to segment. It is also used in unit tests to mock and
+ * spy on the methods to ensure proper behavior
+ * @param {number} flushAt - number of events to queue before sending to segment
+ * @param {nmber} flushInterval - ms interval to flush queue and send to segment
```
Comment (brad-decker) [inline shared/constants/metametrics.js:109]: typo.

File: test/unit/app/controllers/metametrics-test.js [hunk 8]
```diff
@@ -0,0 +1,505 @@
...
+
+function getMockNetworkController(
+  chainId = FAKE_CHAIN_ID,
+  provider = { type: NETWORK },
+) {
+  let networkStore = { chainId, provider }
+  const on = sinon.stub().withArgs('networkDidChange')
+  const updateState = (newState) => {
+    networkStore = { ...networkStore, ...newState }
+    console.log(networkStore)
```
Comment (brad-decker) [inline test/unit/app/controllers/metametrics-test.js:55]: missed a console.log

**Comment by brad-decker** [comment]
@Gudahtt this is ready now, I updated it to move the state from preferences controller.

File: app/scripts/controllers/metametrics.js [hunk 9]
```diff
@@ -0,0 +1,415 @@
...
+  get metaMetricsSendCount() {
+    return this.store.getState().metaMetricsSendCount
+  }
+
+  get metaMetricsId() {
+    return this.store.getState().metaMetricsId
+  }
+
+  set metaMetricsId(metaMetricsId) {
+    this.store.updateState({ metaMetricsId })
```
Comment (Gudahtt) [inline app/scripts/controllers/metametrics.js:193]: Oh! I was going to say, why is `metaMetricsId` being set as an instance field when it's in state? But it's not.

Do you find that using a setter like this is clearer than calling `updateState`? I found it to be a bit misleading.

File: app/scripts/controllers/metametrics.js [hunk 10]
```diff
@@ -0,0 +1,415 @@
...
+   *  track new schema events to
+   * @param {string} segmentLegacyWriteKey - the write key for the Segment
+   *  source to tack legacy events to
+   * @param {string} segmentHost - the host to send events to. The default
+   *  host for Analytics constructor is 'https://api.segment.io'
+   * @param {number} flushAt - the maximum number of events to queue before
+   *  sending to Segment
+   * @param {number} flushInterval - the maximum number of milliseconds to wait
+   *  for before sending queued events
+   * @param {Object} preferencesController - The preferences controller, used
```
Comment (Gudahtt) [inline app/scripts/controllers/metametrics.js:77]: One of the suggestions in the background controller refactor proposal was to limit what we pass into controllers to just what's required. So passing in just the store, or just the methods required, rather than the entire controller. The thought was that this would make the relationships between controllers more obvious, and would make unit tests easier to write because the mocks would be simpler.

File: shared/constants/metametrics.js [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+ * @property {object}  [sensitiveProperties] - Object of sensitive values to
+ *  track. Keys in this object must be in snake_case. These properties will be
+ *  sent in an additional event that excludes the user's metaMetricsId
+ * @property {number}  [revenue] - amount of currency that event creates in
+ *  revenue for MetaMask
+ * @property {string}  [currency] - ISO 4127 format currency for events with
+ *  revenue, defaults to US dollars
+ * @property {number}  [value] - Abstract business "value" attributable to
+ *  customers who trigger this event
+ * @property {MetaMetricsDynamicContext} [eventContext] - additional context
```
Comment (Gudahtt) [inline shared/constants/metametrics.js:59]: Is there still a `typedef` for `MetaMetricsDynamicContext` somewhere? I saw one deleted in `shared/modules/metametrics.js` but I can't find one added.

File: shared/constants/metametrics.js [hunk 12]
```diff
@@ -0,0 +1,213 @@
...
+ * @property {string} app.version - the version of the application
+ * @property {string} userAgent - the useragent string of the user
+ * @property {MetaMetricsPageObject} [page] - an object representing details of
+ *  the current page
+ * @property {MetaMetricsReferrerObject} [referrer] - for metamask, this is the
+ *  dapp that triggered an interaction
+ */
+
+/**
+ * @typedef {Object} MetaMetricsEventPayload
```
Comment (Gudahtt) [inline shared/constants/metametrics.js:43]: Does it still make sense for these type definitions to live here, rather than alongside the function they're used with? I'm not sure I see the benefit - I found myself switching back and forth a lot to try and match things up.

**Review by Gudahtt — COMMENTED** [review]
I'm about halfway through reviewing this - left some initial comments.

I'm really looking forward to getting rid of `getTrackMetaMetricsEvent`! I always found the factory to be a bit convoluted and hard to follow.

File: shared/constants/metametrics.js (hunk 12 above)
Comment (brad-decker) [reply shared/constants/metametrics.js:43]: I use these types in the frontend as well to type the methods that are used for tracking the events through the background, so at least to me, it makes sense to store them in a shared location. Though I guess there is an argument to be made for "storing them in shared/ means you always have to flip files to look them up, instead of importing them directly from the MetaMetrics controller file where you'd only have to look them up from the frontend".

I honestly flip-flopped on this the entire time I was writing them.

File: shared/constants/metametrics.js (hunk 11 above)
Comment (brad-decker) [reply shared/constants/metametrics.js:59]: Nope, this property no longer exists and instead [page] and [referrer] are top-level properties that apparently never got updated. I really want to figure out how to make typescript check our JSdocs only and like, ignore all of the other untyped code <_< (until we get full support obviously)

File: app/scripts/controllers/metametrics.js [hunk 13]
```diff
@@ -0,0 +1,415 @@
...
+    } else if (bool === false) {
+      this.metaMetricsId = null
+    }
+    return this.metaMetricsId
+  }
+
+  /**
+   * retrieve the user's preference for participating in MetaMetrics
+   */
+  get participateInMetaMetrics() {
```
[the remark this answers is not in the document]
Comment (brad-decker) [reply app/scripts/controllers/metametrics.js:176]: I wrote these lines when I first started writing the controller and gradually became more comfortable with the patterns. When I wrote them i was attaching the `preferencesStore.store.getState` method a private method `_getPreferences`. It just seemed cleaner to write `this.participateInMetaMetrics` instead of calling that method.

In the [words of the villager ](https://i.stack.imgur.com/VtefW.gif) from Monty Python and the holy grail:
> I got better...

File: app/scripts/controllers/metametrics.js [hunk 14]
```diff
@@ -0,0 +1,415 @@
...
+   */
+  constructor({
+    isDevOrTestEnvironment,
+    segmentMock,
+    segmentLegacyMock,
+    segmentWriteKey,
+    segmentLegacyWriteKey,
+    segmentHost,
+    flushAt,
+    flushInterval,
```
[the remark this answers is not in the document]
Comment (brad-decker) [reply app/scripts/controllers/metametrics.js:94]: Yep! In the shared controller proposal, I suggest that we drop the usage of `analytics-node` and `analytics-react-native` and instead use the segment API to send events. This would allow for a truly shared implementation that aligns across both platforms. Still, I'd like us to explicitly pass the fetch implementation in. Even if this were to land in extension your suggestion makes a ton of sense.

File: app/scripts/controllers/metametrics.js (hunk 10 above)
Comment (brad-decker) [reply app/scripts/controllers/metametrics.js:77]: Agree.

File: app/scripts/controllers/metametrics.js [hunk 15]
```diff
@@ -0,0 +1,415 @@
...
+      if (flushImmediately) {
+        target.flush()
+      }
+    })
+  }
+
+  /**
+   * track a page view with Segment
+   * @param {MetaMetricsPagePayload} payload - details of the page viewed
+   * @param {MetaMetricsPageOptions} options - options for handling the page
```
[the remark this answers is not in the document]
Comment (brad-decker) [reply app/scripts/controllers/metametrics.js:333]: It is, because the only key in the PageOptions at present is optional.

File: app/scripts/controllers/metametrics.js [hunk 16]
```diff
@@ -0,0 +1,415 @@
...
+      referrer,
+    }
+  }
+
+  /**
+   * Build's the event payload, processing all fields into a format that can be
+   * fed to Segment's track method
+   * @private
+   * @param {MetaMetricsEventPayload} rawPayload - raw payload provided to trackEvent
+   * @param {MetaMetricsEventOptions} options - options for handling/routing event
```
[the remark this answers is not in the document]
Comment (brad-decker) [reply app/scripts/controllers/metametrics.js:222]: It does, it is imported/declared in the block of imports at the beginning of the file.

File: app/scripts/controllers/metametrics.js [hunk 17]
```diff
@@ -0,0 +1,415 @@
...
+    preferencesController,
+    networkController,
+    version,
+    environment,
+    initState,
+  }) {
+    const prefState = preferencesController.store.getState()
+    this.chainId = networkController.getCurrentChainId()
+    this.network = networkController.getNetworkIdentifier()
+    this.locale = prefState.currentLocale?.replace('_', '-') ?? ''
```
[the remark this answers is not in the document]
Comment (brad-decker) [reply app/scripts/controllers/metametrics.js]: resolved in https://github.com/MetaMask/metamask-extension/pull/9857/commits/673a6508e7dc7c53cb04e6cd38829da33aae15d1

File: app/scripts/controllers/metametrics.js [hunk 18]
```diff
@@ -0,0 +1,415 @@
...
+
+    this.store = new ObservableStore({
+      participateInMetaMetrics: null,
+      metaMetricsId: null,
+      metaMetricsSendCount: 0,
+      ...initState,
+    })
+
+    preferencesController.store.subscribe(({ currentLocale }) => {
+      this.locale = currentLocale?.replace('_', '-') ?? ''
```
[the remark this answers is not in the document]
Comment (brad-decker) [reply app/scripts/controllers/metametrics.js]: Resolved in https://github.com/MetaMask/metamask-extension/pull/9857/commits/673a6508e7dc7c53cb04e6cd38829da33aae15d1

File: app/scripts/controllers/metametrics.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (brad-decker) [reply app/scripts/controllers/metametrics.js]: resolved in https://github.com/MetaMask/metamask-extension/pull/9857/commits/673a6508e7dc7c53cb04e6cd38829da33aae15d1

File: app/scripts/controllers/metametrics.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (brad-decker) [reply app/scripts/controllers/metametrics.js]: resolved in https://github.com/MetaMask/metamask-extension/pull/9857/commits/673a6508e7dc7c53cb04e6cd38829da33aae15d1

File: app/scripts/controllers/metametrics.js (hunk 9 above)
Comment (brad-decker) [reply app/scripts/controllers/metametrics.js]: removed in favor of calling updateState directly in https://github.com/MetaMask/metamask-extension/pull/9857/commits/673a6508e7dc7c53cb04e6cd38829da33aae15d1

File: shared/constants/metametrics.js (hunk 11 above)
Comment (brad-decker) [reply shared/constants/metametrics.js]: resolved in https://github.com/MetaMask/metamask-extension/pull/9857/commits/673a6508e7dc7c53cb04e6cd38829da33aae15d1

File: app/scripts/controllers/metametrics.js (hunk 10 above)
Comment (brad-decker) [reply app/scripts/controllers/metametrics.js:58]: Q: Changing this to just pass in preferences store is easy, but doing the same for the network controller would I need to pass in the 'on' method as well as the various getters so that I can subscribe to the network did change event? @Gudahtt

File: app/scripts/controllers/metametrics.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (Gudahtt) [reply app/scripts/controllers/metametrics.js:164]: Sorry I meant the _parameter_ (`options`), not the type. There is no `options` parameter.

File: app/scripts/controllers/metametrics.js (hunk 17 above)
[the remark this answers is not in the document]
Comment (brad-decker) [reply app/scripts/controllers/metametrics.js]: I actually remember why I did this now... The failing tests from `metamask-controller-test.js` -- it seems to not be defined in that test, but that doesn't mean it is undefined in a normal first-time run. Should I update the `initState` of MetaMask Controller in the test?

File: app/scripts/controllers/metametrics.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (brad-decker) [reply app/scripts/controllers/metametrics.js:164]: Sorry, good call. Removed in latest commit

File: app/scripts/controllers/metametrics.js (hunk 17 above)
[the remark this answers is not in the document]
Comment (Gudahtt) [reply app/scripts/controllers/metametrics.js]: Oh, good find! It looks like the controller is missing the `initLangCode` option. It is always passed in as a constructor option in the background.

File: app/scripts/controllers/metametrics.js [hunk 19]
```diff
@@ -0,0 +1,357 @@
...
+   *  to access and subscribe to preferences that will be attached to events
+   * @param {Object} networkController - The network controller, used to access
+   *  the current network's chainId and network identifier. These properties
+   *  are attached to events
+   * @param {string} version - The version of the extension
+   * @param {string} environment - The environment the extension is running in
+   * @param {MetaMetricsControllerState} - State to initialized with
+   */
+  constructor({
+    segment,
```
Comment (Gudahtt) [inline app/scripts/controllers/metametrics.js:68]: These two options appear to be missing from the JSDoc comment

File: app/scripts/controllers/metametrics.js (hunk 10 above)
Comment (Gudahtt) [reply app/scripts/controllers/metametrics.js:58]: We dealt with a similar case here by early-binding `on` to the event we wanted to subscribe to: https://github.com/MetaMask/metamask-extension/blob/develop/app/scripts/metamask-controller.js#L149

To avoid passing in the getters separately, we could setup the subscription entirely in `metamask-controller`, and pass in a handler that gets the `chainId` and network identifier directly :thinking: 

Not sure what I think about that. I wish we could use the store for all of our getters (via selector-like functions), but that might be awkward as things stand currently. Passing in the getters and subscribe function separately seems OK too - whatever works.

File: shared/constants/metametrics.js (hunk 12 above)
Comment (Gudahtt) [reply shared/constants/metametrics.js:43]: Ah I see - I mistakenly thought this type was only used in the background. I see it's used in the context as well. Shared makes perfect sense then.

File: app/scripts/controllers/metametrics.js (hunk 10 above)
Comment (brad-decker) [reply app/scripts/controllers/metametrics.js]: I went with the easier approach of passing in the methods and subscribe method for now.

File: app/scripts/controllers/metametrics.js (hunk 17 above)
[the remark this answers is not in the document]
Comment (brad-decker) [reply app/scripts/controllers/metametrics.js]: I added the initLangCode to the two tests that were missing it (metamask-controller-test, and ui/actions.spec)

File: test/unit/app/controllers/metametrics-test.js [hunk 20]
```diff
@@ -0,0 +1,546 @@
...
+import MetaMetricsController from '../../../../app/scripts/controllers/metametrics'
+import { ENVIRONMENT_TYPE_BACKGROUND } from '../../../../app/scripts/lib/enums'
+import { createSegmentMock } from '../../../../app/scripts/lib/segment'
+import {
+  METAMETRICS_ANONYMOUS_ID,
+  METAMETRICS_BACKGROUND_PAGE_OBJECT,
+} from '../../../../shared/constants/metametrics'
+import waitUntilCalled from '../../../lib/wait-until-called'
+
+const segment = createSegmentMock(2, 10000)
```
Comment (Gudahtt) [inline test/unit/app/controllers/metametrics-test.js:12]: It looks like these two are being used as shared mocks. It would be better to construct these per-test, so that there can be no chance of tests affecting each other.

Comment (Gudahtt) [reply test/unit/app/controllers/metametrics-test.js:12]: Oh I see, it's being flushed after each test :thinking:

File: ui/app/contexts/metametrics.new.js [hunk 21]
```diff
@@ -4,33 +4,49 @@
...
+/**
+ * @typedef {import('../../../shared/constants/metametrics').MetaMetricsEventPayload} MetaMetricsEventPayload
+ * @typedef {import('../../../shared/constants/metametrics').MetaMetricsEventOptions} MetaMetricsEventOptions
+ * @typedef {import('../../../shared/constants/metametrics').MetaMetricsPageObject} MetaMetricsPageObject
+ * @typedef {import('../../../shared/constants/metametrics').MetaMetricsReferrerObject} MetaMetricsReferrerObject
+ */
+
+// types
+/**
+ * @typedef {Omit<MetaMetricsEventPayload, 'environmentType' | 'context'>} UIMetricsEventPayload
```
Comment (Gudahtt) [inline ui/app/contexts/metametrics.new.js:35]: I don't think `context` is a property on `MetaMetricsEventPayload`. Is this what you meant?

```suggestion
 * @typedef {Omit<MetaMetricsEventPayload, 'environmentType' | 'page' | 'referrer'>} UIMetricsEventPayload
```

File: ui/app/contexts/metametrics.new.js [hunk 22]
```diff
@@ -4,33 +4,49 @@
...
+ */
+/**
+ * @typedef {Omit<MetaMetricsEventOptions, 'metaMetricsSendCount'>} UIMetricsEventOptions
```
Comment (Gudahtt) [inline ui/app/contexts/metametrics.new.js:38]: There is no `metaMetricsSendCount` property in the `MetaMetricsEventOptions` type:
```suggestion
 * @typedef {MetaMetricsEventOptions} UIMetricsEventOptions
```

File: ui/app/contexts/metametrics.new.js [hunk 23]
```diff
@@ -128,72 +127,53 @@ export function MetaMetricsProvider({ children }) {
...
-    location,
-    version,
-    locale,
-    context,
-    network,
-    metaMetricsId,
-    participateInMetaMetrics,
-  ])
+    previousMatch.current = match?.path
+  }, [location, dispatch, context])
```
Comment (Gudahtt) [inline ui/app/contexts/metametrics.new.js:176]: I can't see where `dispatch` is used :thinking:

File: ui/app/contexts/metametrics.new.js [hunk 24]
```diff
@@ -128,72 +127,53 @@ export function MetaMetricsProvider({ children }) {
...
+      exact: true,
+      strict: true,
+    })
+    // Start by checking for a missing match route. If this falls through to the else if, then we know we
+    // have a matched route for tracking.
+    if (!match) {
+      // We have more specific pages for each type of transaction confirmation
+      // The user lands on /confirm-transaction first, then is redirected based on
+      // the contents of state.
+      if (location.pathname !== '/confirm-transaction') {
```
Comment (Gudahtt) [inline ui/app/contexts/metametrics.new.js:142]: Wasn't this case removed in a previous PR? I thought `confirm-transaction` was added to the list of paths. Maybe this was reintroduced during a rebase accidentally?

Comment (brad-decker) [reply ui/app/contexts/metametrics.new.js]: Thanks, restored the prior fix.

File: ui/app/contexts/metametrics.new.js (hunk 23 above)
Comment (brad-decker) [reply ui/app/contexts/metametrics.new.js]: Left over from when this logic was in that thunk. removed!

File: ui/app/contexts/metametrics.new.js (hunk 22 above)
Comment (brad-decker) [reply ui/app/contexts/metametrics.new.js]: Thanks!

File: ui/app/contexts/metametrics.new.js (hunk 21 above)
Comment (brad-decker) [reply ui/app/contexts/metametrics.new.js]: absolutely. added in one of the latest commits

File: app/scripts/metamask-controller.js [hunk 25]
```diff
@@ -298,9 +287,9 @@ export default class MetamaskController extends EventEmitter {
       ),
       provider: this.provider,
       blockTracker: this.blockTracker,
-      trackMetaMetricsEvent: this.trackMetaMetricsEvent,
+      trackMetaMetricsEvent: this.metaMetricsController.trackEvent,
       getParticipateInMetrics: () =>
-        this.preferencesController.getParticipateInMetaMetrics(),
+        this.metaMetricsController.participateInMetaMetrics,
```
[the remark this answers is not in the document]
Comment (brad-decker) [reply app/scripts/metamask-controller.js]: Yep, added in latest commits

File: app/scripts/controllers/metametrics.js [hunk 26]
```diff
@@ -0,0 +1,366 @@
...
+      return
+    }
+    const { metaMetricsId } = this.state
+    const idTrait = metaMetricsId ? 'userId' : 'anonymousId'
+    const idValue = metaMetricsId ?? METAMETRICS_ANONYMOUS_ID
+    this.segment.page({
+      [idTrait]: idValue,
+      name,
+      properties: {
+        // We do not want to send addresses or accounts in any events
```
[the remark this answers is not in the document]
Comment (brad-decker) [reply app/scripts/controllers/metametrics.js]: Returned to its prior location, where the omit call is made

File: app/scripts/metamask-controller.js [hunk 27]
```diff
@@ -362,6 +351,7 @@ export default class MetamaskController extends EventEmitter {
       TransactionController: this.txController.store,
       KeyringController: this.keyringController.store,
       PreferencesController: this.preferencesController.store,
+      MetamaskController: this.metaMetricsController.store,
```
[the remark this answers is not in the document]
Comment (brad-decker) [reply app/scripts/metamask-controller.js]: <_< oops

File: app/scripts/lib/segment.js [hunk 28]
```diff
@@ -0,0 +1,101 @@
+import Analytics from 'analytics-node'
+
+const isDevOrTestEnvironment = Boolean(
+  process.env.METAMASK_DEBUG || process.env.IN_TEST,
+)
+const { SEGMENT_WRITE_KEY } = process.env
```
[the remark this answers is not in the document]
Comment (brad-decker) [reply app/scripts/lib/segment.js]: I used nullish coalescing to prevent eslint from automatically destructuring these. You are correct that that doesn't work. I also tried just using `process.env.SEGMENT_WRITE_KEY` inline but this made the ternaries later in the file unwieldy.

File: app/scripts/controllers/metametrics.js [hunk 29]
```diff
@@ -0,0 +1,366 @@
...
+      page,
+      referrer,
+    }
+  }
+
+  /**
+   * Build's the event payload, processing all fields into a format that can be
+   * fed to Segment's track method
+   * @private
+   * @param {MetaMetricsEventPayload} rawPayload - raw payload provided to trackEvent
```
[the remark this answers is not in the document]
Comment (brad-decker) [reply app/scripts/controllers/metametrics.js]: Thanks! Added in latest commits

File: app/scripts/lib/segment.js (hunk 28 above)
[the remark this answers is not in the document]
Comment (Gudahtt) [reply app/scripts/lib/segment.js]: Huh, interesting solution! I'll have to remember that.

**Review by Gudahtt — DISMISSED** [review]
LGTM!

**Comment by Gudahtt** [comment]
This is a fantastic reorganization by the way. The complexity of our handling of metrics  has been going up a fair bit lately, but this seems like a substantial step forward. It may even be simpler now than it was prior to Segment, at least if you exclude the legacy events. It at least makes a lot more sense to me. In particular I love that we're not using a factory for the track event anymore.

**Comment by Gudahtt** [comment]
I'll double-check locally that the migration works correctly. I'm also looking into this metrics e2e test failure.

**Comment by Gudahtt** [comment]
I've pushed a commit temporarily to remove a less-than-informative assertion from the failing test. It is removed in PR #9974 as well.

**Comment by Gudahtt** [comment]
Welp. I got the previous commit to pass! Must be an intermittent failure - possibly pre-existing. Nothing in this PR stands out as a likely cause :thinking:

**Review by Gudahtt — APPROVED** [review]
I have tested the migration locally, and it all seems to work correctly.
